### PR TITLE
Update nodesdk and add resource detectors

### LINKED ISSUE #1113: Update NodeSDK to 0.33 and add resource detectors
The NodeSDK now allows to have additional resource detectors being loaded on startup: 

https://github.com/open-telemetry/opentelemetry-js/releases/tag/experimental%2Fv0.33.0
&
https://deploy-preview-1767--opentelemetry.netlify.app/docs/instrumentation/js/resources/ (WIP)

The auto instrumentation of the operator could make use of that and auto-detect container resources (and [more](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/detectors/node) if running in a cloud environment)

What needs to be done:
* Update dependencies in [package.json](https://github.com/open-telemetry/opentelemetry-operator/blob/main/autoinstrumentation/nodejs/package.json) to 0.33 & add additional dependencies
* Update [autoinstrumentation.ts](https://github.com/open-telemetry/opentelemetry-operator/blob/main/autoinstrumentation/nodejs/src/autoinstrumentation.ts) to use additional resource detectors:

```js
const sdk = new NodeSDK({
    autoDetectResources: true,
    instrumentations: [getNodeAutoInstrumentations()],
    traceExporter: new OTLPTraceExporter(),
    resourceDetectors: [envDetector, processDetector, dockerCGroupV1Detector, ...],
});
```

cc: @abhee11, @osherv

### PR DESCRIPTION
**Pull request by sairam866**
Fixes #1113

### CODE REVIEW

**Comment by pavolloffay** [comment]
@svrnm could you please review it?

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts [hunk 1]
```diff
@@ -1,12 +1,48 @@
...
 import { NodeSDK } from '@opentelemetry/sdk-node';
 
 const sdk = new NodeSDK({
     autoDetectResources: true,
     instrumentations: [getNodeAutoInstrumentations()],
     traceExporter: new OTLPTraceExporter(),
+    resourceDetectors:
+        [
+            // Standard resource detectors.
+            browserDetector,
```
Comment (svrnm) [inline autoinstrumentation/nodejs/src/autoinstrumentation.ts:26]:
```suggestion
```

This is only for browser environments

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts [hunk 2]
```diff
@@ -1,12 +1,48 @@
...
+            awsBeanstalkDetector,
+            awsEc2Detector,
+            awsEcsDetector,
+            awsEksDetector,
+            awsLambdaDetector,
+            gcpDetector,
+            gitHubDetector,
+
+            // Agent resource detectors.
+            instanaAgentDetector,
```
Comment (svrnm) [inline autoinstrumentation/nodejs/src/autoinstrumentation.ts:44]: I could only verify that they fail (almost) silently if none of those is available, I don't know how they behave in certain setups (what does the gcpDetector do if run in an aws cloud etc.)

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts [hunk 3]
```diff
@@ -1,12 +1,48 @@
...
+            containerDetector,
+            envDetector,
+            hostDetector,
+            osDetector,
+            processDetector,
+
+            // Cloud resource detectors.
+            alibabaCloudEcsDetector,
+            awsBeanstalkDetector,
+            awsEc2Detector,
```
Comment (svrnm) [inline autoinstrumentation/nodejs/src/autoinstrumentation.ts:36]: is this even applicable in k8s?

**Comment by svrnm** [comment]
> @svrnm could you please review it?

Overall it looks good to me, provided a few comments

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts (hunk 1 above)
Comment (sairam866) [reply autoinstrumentation/nodejs/src/autoinstrumentation.ts]: Removed.

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts (hunk 2 above)
Comment (sairam866) [reply autoinstrumentation/nodejs/src/autoinstrumentation.ts]: The resource detectors hit the cloud provider's endpoint and times out when not in proper environment. 
If so, the logic fails silently with empty resource from that detector.

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts (hunk 3 above)
Comment (sairam866) [reply autoinstrumentation/nodejs/src/autoinstrumentation.ts]: Realized it and removed not applicable detectors.

**Comment by sairam866** [comment]
@pavolloffay Can you enable the workflow for me (a first time committer) and review the PR ?

File: autoinstrumentation/nodejs/src/autoinstrumentation.ts [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
 import { NodeSDK } from '@opentelemetry/sdk-node';
 
 const sdk = new NodeSDK({
     autoDetectResources: true,
     instrumentations: [getNodeAutoInstrumentations()],
     traceExporter: new OTLPTraceExporter(),
+    resourceDetectors:
+        [
+            // Standard resource detectors.
+            containerDetector,
```
Comment (pavolloffay) [inline autoinstrumentation/nodejs/src/autoinstrumentation.ts:18]: Do you have a span example e.g. from Jaeger that shows which resource attributes are collected? 

Note that the operator as well injects the env var OTEL_RESOURCE_ATTRIBUTES with k8s attributes (container name, pod etc.)

Comment (pavolloffay) [reply autoinstrumentation/nodejs/src/autoinstrumentation.ts:18]: > Does this work all the time or only in Sidecar mode (as the README.md suggests)?

The OTEL_RESOURCE_ATTRIBUTES is injected into all application containers that use the instrumentation CR.

The collector CR injects the attributes as well when deployed in the sidecar mode

Comment (svrnm) [reply autoinstrumentation/nodejs/src/autoinstrumentation.ts:18]: > > Does this work all the time or only in Sidecar mode (as the README.md suggests)?
> 
> The OTEL_RESOURCE_ATTRIBUTES is injected into all application containers that use the instrumentation CR.
> 
> The collector CR injects the attributes as well when deployed in the sidecar mode

That's interesting, is there a full list of attributes that are injected (somewhere in the source code I assume?)

File: autoinstrumentation/nodejs/package.json [hunk 5]
```diff
@@ -14,9 +14,14 @@
...
-    "@opentelemetry/api": "1.1.0",
-    "@opentelemetry/auto-instrumentations-node": "0.31.0",
-    "@opentelemetry/exporter-trace-otlp-grpc": "0.31.0",
-    "@opentelemetry/sdk-node": "0.31.0"
+    "@opentelemetry/api": "1.2.0",
+    "@opentelemetry/auto-instrumentations-node": "0.33.1",
+    "@opentelemetry/exporter-trace-otlp-grpc": "0.33.0",
+    "@opentelemetry/resource-detector-alibaba-cloud": "0.27.2",
+    "@opentelemetry/resource-detector-aws": "1.1.2",
+    "@opentelemetry/resource-detector-container": "0.1.2",
```
Comment (svrnm) [inline autoinstrumentation/nodejs/package.json:22]: can you make use of the latest version (0.2.0), this contains an update for containers running on cgroupv2

**Review by svrnm — APPROVED** [review]
[Minor improvement](https://github.com/open-telemetry/opentelemetry-operator/pull/1211/files#r1018358180), which could also be a fast follow up, overall it now looks good to me.

**Comment by svrnm** [comment]
@pavolloffay any reasons not to merge this yet? What's missing from your point of view?

**Review by pavolloffay — APPROVED** [review]
LGTM, but the PR needs to be rebased

**Comment by sairam866** [comment]
@pavolloffay @svrnm - Updated the PR with rebase and latest versions.

**Review by svrnm — APPROVED** [review]
still lgtm

**Comment by sairam866** [comment]
@pavolloffay @svrnm - Looks like the unit test is failing randomly and not caused by this PR.

**Pull request merged**